sql: set zone configs before backfill on ALTER TABLE LOCALITY/PRIMARY KEY

otan opened the pull request:
Ensure that the zone configurations are correctly set before the
backfill runs, so data is backfilled to the correct location upfront for
REGIONAL BY ROW tables. Note this bug also exists on existing ALTER
PRIMARY KEY statements.

Release justification: bug fixes and low-risk updates to new
functionality

Release note (bug fix): Fix bug where zone configurations on indexes
were not copied before the backfill of an ALTER PRIMARY KEY. They used
to be copied afterwards instead.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61300)
<!-- Reviewable:end -->

otan commented on pkg/sql/schema_changer.go:
```diff
@@ -1118,75 +1137,15 @@ func (sc *SchemaChanger) done(ctx context.Context) error {
 
 			// Some primary key change specific operations need to happen before
```
what a false claim ;~;

ajstorm reviewed (changes requested):
> Reviewed 2 of 3 files at r1, 4 of 5 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/ccl/multiregionccl/regional_by_row_test.go, line 144 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUogBTB5NhRZFtg4f4Y:-MUogBTB5NhRZFtg4f4Z:b-lp8cqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/ccl/multiregionccl/regional_by_row_test.go#L144)):*
> ```Go
> 	range_max_bytes = 536870912,
> 	gc.ttlseconds = 90000,
> 	num_replicas = 3,
> ```

I might be missing something, but shouldn't this have a `global_reads` value set?  Or are we dropping the zone config first, and then apply the new one at the end?
___
*[pkg/ccl/multiregionccl/regional_by_row_test.go, line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUogcpV8E00C5G-HpAS:-MUogcpV8E00C5G-HpAT:bvexbh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/ccl/multiregionccl/regional_by_row_test.go#L187)):*
> ```Go
> 	const showCreateTableStringSQL = `SELECT create_statement FROM [SHOW CREATE TABLE t.test]`
> 	const zoneConfigureSQLStatements = `
> 		SELECT coalesce(string_agg(raw_config_sql, ';' ORDER BY raw_config_sql), 'NULL')
> ```

Why not ```SELECT zone_config, index_name, partition_name FROM [SHOW PARTITIONS FROM TABLE test];```
___
*[pkg/sql/schema_changer.go, line 2428 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpEuPs7ZX2QeC77b1R:-MUpEuPtDIj6xrV_T1v7:brpwdej) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/schema_changer.go#L2428)):*
> ```Go
> ) error {
> 	if pkSwap := mutation.GetPrimaryKeySwap(); pkSwap != nil {
> 		// We might have to update some zone configs for indexes that are
> ```

Nit: Does this comment belong in with the caller?
___
*[pkg/sql/schema_changer.go, line 2455 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpDw_OCqg9YbJ02QFh:-MUpDw_OCqg9YbJ02QFi:b-urcwk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/schema_changer.go#L2455)):*
> ```Go
> 				case *descpb.TableDescriptor_LocalityConfig_Global_,
> 					*descpb.TableDescriptor_LocalityConfig_RegionalByTable_:
> 					// Just the table re-writing the locality config change will suffice.
> ```

Nit: wording - "Nothing to do here. The table re-writing the locality config is all that's required"?
___
*[pkg/sql/schema_changer.go, line 2492 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpEz6a5mj0JBvvP7s6:-MUpEz6a5mj0JBvvP7s7:bn80h25) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/schema_changer.go#L2492)):*
> ```Go
> 			)
> 		}
> 		// For the normal case, copy the zone configs over.
> ```

Nit: Can we explain what we mean by "normal" here?
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpnsSN-UE8UE8A06eC:-MUpnsSN-UE8UE8A06eD:b-6v9eb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L31)):*
> ```Go
> // AlterPrimaryKeyCorrectZoneConfigIntermediateZoneConfig is an expected
> // intermediate zone configuration in the AlterPrimaryKeyCorrectZoneConfigTestCase.
> type AlterPrimaryKeyCorrectZoneConfigIntermediateZoneConfig struct {
> ```

Nit: these names are very long.  Can we not put this struct in `alter_primary_key_test.go`, give it a shorter name, and stop using it in `regional_by_row_test.go` (by moving that test into `alter_primary_key_test.go`)?  
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 54 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUppPlR0d_FKo6UREcq:-MUppPlR0d_FKo6UREcr:btr9wwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L54)):*
> ```Go
> 	defer SetTestJobsAdoptInterval()()
> 
> 	var chunkSize int64 = 100
> ```

Nit: ƒa`:=` for these two?
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpr9kU0GXxlG7LmJ7g:-MUpr9kU0GXxlG7LmJ7h:bnh0lru) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L103)):*
> ```Go
> 			// Disable strict GC TTL enforcement because we're going to shove a zero-value
> 			// TTL into the system with AddImmediateGCZoneConfig.
> 			defer DisableGCTTLStrictEnforcement(t, sqlDB)()
> ```

Nit: where is the call to  AddImmediateGCZoneConfig?
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpriy_4GkwWJCxEm4x:-MUpriy_4GkwWJCxEm4y:bmtfrqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L107)):*
> ```Go
> 			if _, err := sqlDB.Exec(fmt.Sprintf(`
> %s;
> USE t;
> ```

Nit: are we assuming that the database here is named `t`?  If so, is this a reasonable assumption?
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpsN969oXOyT837xxX:-MUpsN969oXOyT837xxY:b-1nkqbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L110)):*
> ```Go
> %s
> `, createDBStatement, tc.SetupQuery)); err != nil {
> 				t.Fatal(err)
> ```

Nit: Why `Fatal` vs. `require.NoError`?
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpsY1o2I4U5-7LiHOx:-MUpsY1o2I4U5-7LiHOy:bhvvm3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L113)):*
> ```Go
> 			}
> 
> 			if err := BulkInsertIntoTable(sqlDB, maxValue); err != nil {
> ```

Nit: Can you add a comment here as to why you're performing this bulk insert?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @arulajmani)

___
*[pkg/ccl/multiregionccl/regional_by_row_test.go, line 144 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUogBTB5NhRZFtg4f4Y:-MUq4jCk2wrz_IBA40G6:b6i0dgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/ccl/multiregionccl/regional_by_row_test.go#L144)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

I might be missing something, but shouldn't this have a `global_reads` value set?  Or are we dropping the zone config first, and then apply the new one at the end?
</blockquote></details>

i believe we should only apply the table level zone configs once `done()` is called.
however, the *new* indexes (including the new PKs) should be done immediately.
___
*[pkg/ccl/multiregionccl/regional_by_row_test.go, line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUogcpV8E00C5G-HpAS:-MUq4xcYAaWt-qoJ5wH1:b5gwkkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/ccl/multiregionccl/regional_by_row_test.go#L187)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Why not ```SELECT zone_config, index_name, partition_name FROM [SHOW PARTITIONS FROM TABLE test];```
</blockquote></details>

because comparing individual rows is a lot more effort. this should give the exact same set of results, as partition_name and index_name is encoded within the raw_config_sql.
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpnsSN-UE8UE8A06eC:-MUq6nAbAjOnMkf9VLMb:bonmrfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L31)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: these names are very long.  Can we not put this struct in `alter_primary_key_test.go`, give it a shorter name, and stop using it in `regional_by_row_test.go` (by moving that test into `alter_primary_key_test.go`)?  
</blockquote></details>

this does not really help with the fact that it would need to be in two separate packages.
this can only be tested in ALTER PRIMARY KEY  where there are partitions. i don't think the test itself belongs in ccl/multiregionccl
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 54 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUppPlR0d_FKo6UREcq:-MUq6xvs37LZZtz0P4lV:b32wzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L54)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: ƒa`:=` for these two?
</blockquote></details>

wa
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpr9kU0GXxlG7LmJ7g:-MUq7Ytd4C_V2I8QXHI0:blqfaa9) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L103)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: where is the call to  AddImmediateGCZoneConfig?
</blockquote></details>

hmm, deleting it. relic of something else.
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpriy_4GkwWJCxEm4x:-MUq7depDncxanFXHRXl:bxzu3gk) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L107)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: are we assuming that the database here is named `t`?  If so, is this a reasonable assumption?
</blockquote></details>

yes, all the helpers here use that database.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/ccl/multiregionccl/regional_by_row_test.go, line 144 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUogBTB5NhRZFtg4f4Y:-MUq8-PL0t_HUwRmkHKF:b85l5lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/ccl/multiregionccl/regional_by_row_test.go#L144)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i believe we should only apply the table level zone configs once `done()` is called.
however, the *new* indexes (including the new PKs) should be done immediately.
</blockquote></details>

Yeah, realized that further on in my review.  All good
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpnsSN-UE8UE8A06eC:-MUq8PVg6mNpFOI-7eSR:bvea15a) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L31)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this does not really help with the fact that it would need to be in two separate packages.
this can only be tested in ALTER PRIMARY KEY  where there are partitions. i don't think the test itself belongs in ccl/multiregionccl
</blockquote></details>

Hmm.  Is this what the @arulajmani sadness feels like?
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 54 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUppPlR0d_FKo6UREcq:-MUq9-0O19yHkEN_V6aq:bu2eh16) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L54)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

wa
</blockquote></details>

Wow, that didn't print out as intended.  My question was, should we be using `:=` for these two?
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpriy_4GkwWJCxEm4x:-MUq9CyF7ZuHha-hLTFM:bli8xgu) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L107)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

yes, all the helpers here use that database.
</blockquote></details>

Seems reasonable for now, but makes this test a bit fragile.  Would be good if we passed in the database name so that we could `USE` the specified database.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
Nice catch! :100: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-:-MUt9jI4C5uJBzgLn7sJ:b-rlg7qf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/ccl/partitionccl/alter_primary_key_test.go, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUt9EIw6U1e7rl1cI6x:-MUt9EIw6U1e7rl1cI6y:b8f9r7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a16fe3c5018870e436f212549efaa74957dff3f/pkg/ccl/partitionccl/alter_primary_key_test.go#L34)):*
> ```Go
> 					ShowConfigStatement: `SHOW ZONE CONFIGURATION FOR INDEX t.test@v_idx_rewrite_for_primary_key_change`,
> 					ExpectedTarget:      `INDEX t.public.test@v_idx_rewrite_for_primary_key_change`,
> 					ExpectedSQL: `ALTER INDEX t.public.test@v_idx_rewrite_for_primary_key_change CONFIGURE ZONE USING
> ```

nit: structuring this whole thing in the form of `SHOW ZONE CONFIGURATION FROM <old_idx_name>` needs to match `SHOW ZONE CONFIGURATION FROM <rewrite_idx_name>` might make this test more readable. 
___
*[pkg/sql/schema_changer.go, line 1138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUk9JVLuziwVb-uPC9p-r1-1138:-MUt7tdP1WL2KAR3_ccv:b8j74ly) ([raw file](https://github.com/cockroachdb/cockroach/blob/82bbb713054a0efd7ef36be63846968e7eb5f360/pkg/sql/schema_changer.go#L1138)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

what a false claim ;~;
</blockquote></details>

Good find! 
___
*[pkg/sql/schema_changer.go, line 2497 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUt8gge8Aq_WNEyvnzH:-MUt8gge8Aq_WNEyvnzI:b-s2p02g) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a16fe3c5018870e436f212549efaa74957dff3f/pkg/sql/schema_changer.go#L2497)):*
> ```Go
> 		// For the plain ALTER PRIMARY KEY case, copy the zone configs over
> 		// for any new indexes.
> 		return maybeUpdateZoneConfigsForPKChange(
> ```

Can we do this only if `isDone` is false? That way we dont have to re-write the same zone configurations after the change has been finalized for regular PK changes. 
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpnsSN-UE8UE8A06eC:-MUt1Ddv4iJtvUKcWZ0m:b6v2f2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L31)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Hmm.  Is this what the @arulajmani sadness feels like?
</blockquote></details>

Any reason not to put this file in CCL (say `pkg/ccl/testutils/<file_name>.go`)? Looks like both its callers are from CCL. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/ccl/partitionccl/alter_primary_key_test.go, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUt9EIw6U1e7rl1cI6x:-MUt_cQiAyiFc_EwIpOO:bsfuoid) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a16fe3c5018870e436f212549efaa74957dff3f/pkg/ccl/partitionccl/alter_primary_key_test.go#L34)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: structuring this whole thing in the form of `SHOW ZONE CONFIGURATION FROM <old_idx_name>` needs to match `SHOW ZONE CONFIGURATION FROM <rewrite_idx_name>` might make this test more readable. 
</blockquote></details>

hmm, the zone configs have different names so doing that regexp makes it equally annoying to read as a gotcha. i think i'll keep this one :P 
___
*[pkg/sql/schema_changer.go, line 2497 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUt8gge8Aq_WNEyvnzH:-MUtc-pJ7tTMkgE-GKdt:b-fs19f1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a16fe3c5018870e436f212549efaa74957dff3f/pkg/sql/schema_changer.go#L2497)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Can we do this only if `isDone` is false? That way we dont have to re-write the same zone configurations after the change has been finalized for regular PK changes. 
</blockquote></details>

hmm, i think so, but to keep it in line with the original comment of the false claim above, i think it's safer to overwrite. i don't think there's harm in re-writing the same zone config is there?
___
*[pkg/sql/sqltestutils/alter_primary_key.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUpnsSN-UE8UE8A06eC:-MUt_nAa9mRXb0wRFUNK:b6q48q4) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44775d4be52ae2a9f2c2cb1fbeb28d3ab6af28/pkg/sql/sqltestutils/alter_primary_key.go#L31)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Any reason not to put this file in CCL (say `pkg/ccl/testutils/<file_name>.go`)? Looks like both its callers are from CCL. 
</blockquote></details>

yeah i'll do that.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/sql/schema_changer.go, line 2497 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/61300#-MUt8gge8Aq_WNEyvnzH:-MUtu4MaBImotZIOo80u:b92vr96) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a16fe3c5018870e436f212549efaa74957dff3f/pkg/sql/schema_changer.go#L2497)):*
> i don't think there's harm in re-writing the same zone config is there?

Naa, I don't think it's that big a deal considering we're writing the same thing again. Would you mind capturing this potential overwrite and that we don't care in a comment here? 


<!-- Sent from Reviewable.io -->

otan committed:
sql: set zone configs before backfill on ALTER TABLE LOCALITY/PRIMARY KEY

Ensure that the zone configurations are correctly set before the
backfill runs, so data is backfilled to the correct location upfront for
REGIONAL BY ROW tables. Note this bug also exists on existing ALTER
PRIMARY KEY statements.

Release justification: bug fixes and low-risk updates to new
functionality

Release note (bug fix): Fix bug where zone configurations on indexes
were not copied before the backfill of an ALTER PRIMARY KEY. They used
to be copied afterwards instead.

otan commented:
bors r=arulajmani,ajstorm
